### PR TITLE
Add Block Transaction Count Endpoint

### PR DESCRIPTION
**Pull request by jelilat**
## Description

Adds `starknet_getBlockTransactionCount` rpc endpoint

## Types of changes

- New feature (non-breaking change which adds functionality)

## Testing

**Requires testing**: Yes

**Did you write tests??**: Yes

### CODE REVIEW

File: rpc/handlers.go [hunk 1]
```diff
@@ -215,3 +215,13 @@ func (h *Handler) GetTransactionByHash(hash *felt.Felt) (*Transaction, *jsonrpc.
...
 }
+
+// https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_api_openrpc.json#L373
+func (h *Handler) GetBlockTransactionCount(id *BlockId) (uint64, *jsonrpc.Error) {
+	block, err := h.getBlockById(id)
+	if err != nil {
+		return 0, ErrBlockNotFound
+	}
+
+	return uint64(len(block.Transactions)), nil
```
Comment (omerfirmak) [inline rpc/handlers.go:226]: We probably dont need the cast to uint64 here

**Review by omerfirmak — APPROVED** [review]

File: rpc/handlers.go (hunk 1 above)
Comment (omerfirmak) [reply rpc/handlers.go:226]: Yeah, I meant we can just return `int`

Comment (jelilat) [reply rpc/handlers.go]: Alright. Done

**Comment by jelilat** [comment]
@omerfirmak made some changes to this cos I don't think we should fetch all the blocks. We won't have to pointlessly fetch block head and receipts, so it should be faster and consume less memory

File: blockchain/blockchain.go [hunk 2]
```diff
@@ -323,6 +339,43 @@ func getBlockByHash(txn db.Transaction, hash *felt.Felt) (block *core.Block, err
...
+	numBytes := make([]byte, lenOfByteSlice)
+	binary.BigEndian.PutUint64(numBytes, number)
+
+	iterator, err := txn.NewIterator()
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		// Prioritise closing error over other errors
+		if closeErr := iterator.Close(); closeErr != nil {
```
Comment (omerfirmak) [inline blockchain/blockchain.go:353]: should use `CloseAndWrapOnError`

File: rpc/handlers.go [hunk 3]
```diff
@@ -215,3 +215,31 @@ func (h *Handler) GetTransactionByHash(hash *felt.Felt) (*Transaction, *jsonrpc.
...
+	} else {
+		count, err = h.bcReader.GetBlockTransactionCountByNumber(id.Number)
+	}
+
+	return count, err
+}
+
+// https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_api_openrpc.json#L373
+func (h *Handler) GetBlockTransactionCount(id *BlockId) (uint64, *jsonrpc.Error) {
+	txnCount, err := h.getBlockTxnCountById(id)
```
Comment (omerfirmak) [inline rpc/handlers.go:239]: no need for the `getBlockTxnCountById` helper

File: rpc/handlers.go [hunk 4]
```diff
@@ -215,3 +215,31 @@ func (h *Handler) GetTransactionByHash(hash *felt.Felt) (*Transaction, *jsonrpc.
...
+	var err error
+	if id.Latest {
+		if height, e := h.bcReader.Height(); err != nil {
+			return 0, e
+		} else {
+			count, err = h.bcReader.GetBlockTransactionCountByNumber(height)
+		}
+	} else if id.Hash != nil {
+		count, err = h.bcReader.GetBlockTransactionCountByHash(id.Hash)
+	} else {
```
Comment (omerfirmak) [inline rpc/handlers.go:230]: id.Pending will hit this branch

File: blockchain/blockchain.go [hunk 5]
```diff
@@ -323,6 +339,43 @@ func getBlockByHash(txn db.Transaction, hash *felt.Felt) (block *core.Block, err
...
+			err = closeErr
+		}
+	}()
+
+	var noOfTransactions uint64
+	prefix := db.TransactionsByBlockNumberAndIndex.Key(numBytes)
+	for iterator.Seek(prefix); iterator.Valid(); iterator.Next() {
```
Comment (omerfirmak) [inline blockchain/blockchain.go:360]: essentially duplicate of `getBlockByNumber`, we should eliminate the duplication

**Review by omerfirmak — CHANGES_REQUESTED** [review]

**Comment by omerfirmak** [comment]
> @omerfirmak made some changes to this cos I don't think we should fetch all the blocks. We won't have to pointlessly fetch block head and receipts, so it should be faster and consume less memory

You still have to iterate over txns, just not unmarshaling them. Iterating over db dominates the runtime anyways, so the benefit we get from this is marginal.

**Comment by IronGauntlets** [comment]
Any particular reason why we are not doing the following:

```go
func (h *Handler) GetBlockTransactionCount(id *BlockId) (uint64, *jsonrpc.Error) {
	block, err := h.getBlockById(id)
	if err != nil {
		return 0, ErrBlockNotFound
	}

	return len(block.Transactions), nil
}
```

I know the above will be wasteful, however, adding `txCount` to the header is out of scope for this PR. 

If we do decide to add `txCount` to the header we should also add `eventCount` along with testcases to ensure there is no mismatch.

**Comment by jelilat** [comment]
> Any particular reason why we are not doing the following:
> 
> ```go
> func (h *Handler) GetBlockTransactionCount(id *BlockId) (uint64, *jsonrpc.Error) {
> 	block, err := h.getBlockById(id)
> 	if err != nil {
> 		return 0, ErrBlockNotFound
> 	}
> 
> 	return len(block.Transactions), nil
> }
> ```
> 
> I know the above will be wasteful, however, adding `txCount` to the header is out of scope for this PR.
> 
> If we do decide to add `txCount` to the header we should also add `eventCount` along with testcases to ensure there is no mismatch.

This is what I did earlier, but I had reservations and made the new changes. If everyone agrees to do it this way, I'll just revert the changes cc @omerfirmak

**Review by IronGauntlets — APPROVED** [review]

**Review by omerfirmak — APPROVED** [review]

**Pull request merged**